pkg/agent/controller/packetsampling/packetin.go

```diff
@@ -0,0 +1,162 @@
+		}
+
+		_, err = c.crdClient.CrdV1alpha1().PacketSamplings().UpdateStatus(context.TODO(), update, v1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("update Traceflow failed: %w", err)
+		}
+		klog.InfoS("Updated PacketSampling", "ps", klog.KObj(ps), "status", update.Status)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("PacketSampling update error: %w", err)
+	}
+
+	if samplingState != nil {
+		rawData := pktIn.Data.(*util.Buffer).Bytes()
+		ci := gopacket.CaptureInfo{
+			Timestamp:     time.Now(),
+			CaptureLength: len(rawData),
+			Length:        len(rawData),
+		}
+		err = samplingState.pcapngWriter.WritePacket(ci, rawData)
+		if err != nil {
+			return fmt.Errorf("couldn't write packet: %w", err)
+		}
+
+		if samplingState.numCapturedPackets == oldPS.Spec.FirstNSamplingConfig.Number && samplingState.shouldSyncPackets {
+			return c.uploadPacketsFile(oldPS)
+		}
+	}
+	return nil
+}
+
+// parsePacketIn parses the packet-in message and returns
+// 1. the sampling state of the PacketSampling (on sampling mode),
+func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (_ *crdv1alpha1.PacketSampling, _ *packetSamplingState, shouldSkip bool, _ error) {
+	// Get data plane tag.
+	// Directly read data plane tag from packet.
+	var err error
+	var tag uint8
+	samplingState := packetSamplingState{}
+
+	etherData := new(protocol.Ethernet)
+	if err := etherData.UnmarshalBinary(pktIn.Data.(*util.Buffer).Bytes()); err != nil {
+		return nil, nil, false, fmt.Errorf("failed to parse Ethernet packet from packet-in message: %v", err)
+	}
+	if etherData.Ethertype == protocol.IPv4_MSG {
+		ipPacket, ok := etherData.Data.(*protocol.IPv4)
+		if !ok {
+			return nil, nil, false, fmt.Errorf("invalid PacketSampling ipv4 packet")
+
+		}
+		tag = ipPacket.DSCP
+	} else if etherData.Ethertype == protocol.IPv6_MSG {
+		ipv6Packet, ok := etherData.Data.(*protocol.IPv6)
+		if !ok {
+			return nil, nil, false, fmt.Errorf("invalid PacketSampling ipv6 packet")
+		}
+		tag = ipv6Packet.TrafficClass >> 2
+	} else {
+		return nil, nil, false, fmt.Errorf("unsupported traceflow packet Ethertype: %d", etherData.Ethertype)
+	}
+
+	c.runningPacketSamplingsMutex.Lock()
+	psState, exists := c.runningPacketSamplings[int8(tag)]
+	if exists {
+		if psState.numCapturedPackets == psState.maxNumCapturedPackets {
+			c.runningPacketSamplingsMutex.Unlock()
+			return nil, nil, true, nil
+		}
+		psState.numCapturedPackets++
+		if psState.numCapturedPackets == psState.maxNumCapturedPackets {
+			c.ofClient.UninstallPacketSamplingFlows(tag)
+		}
+		samplingState = *psState
+
+	}
+	c.runningPacketSamplingsMutex.Unlock()
+	if !exists {
+		return nil, nil, false, fmt.Errorf("PacketSampling for dataplane tag %d not found in cache", tag)
+	}
+
+	ps, err := c.packetSamplingLister.Get(psState.name)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("failed to get PacketSampling %s CRD: %v", psState.name, err)
+	}
```

Comment:
This doesn't seem need to be retrieved and returned by this func.

Reply:
removed.